DeepLinkKit/Router/DPLDeepLinkRouter.m

```diff
@@ -147,45 +143,84 @@ - (BOOL)handleUserActivity:(NSUserActivity *)userActivity withCompletion:(DPLRou
     return NO;
 }
 
+- (UIViewController <DPLTargetViewController> *)viewControllerForUrl:(NSURL *)url {
+    DPLDeepLink *deepLink;
+    id handler;
+    deepLink = [self deepLinkForUrl:url handler:&handler];
+    if (handler) {
+        DPLRouteHandler *routeHandler = [self routeHandlerForHandler:handler];
+        if (routeHandler) {
+            return [self viewControllerForHandler:routeHandler withDeepLink:deepLink];
+        }
+    }
+    return nil;
+}
 
-- (BOOL)handleRoute:(NSString *)route withDeepLink:(DPLDeepLink *)deepLink error:(NSError *__autoreleasing *)error {
-    id handler = self[route];
-    
-    if ([handler isKindOfClass:NSClassFromString(@"NSBlock")]) {
-        DPLRouteHandlerBlock routeHandlerBlock = handler;
-        routeHandlerBlock(deepLink);
+
+- (DPLDeepLink *)deepLinkForUrl:(NSURL *)url handler:(id *)handler {
+    DPLDeepLink *deepLink;
+    for (NSString *route in self.routes) {
+        DPLRouteMatcher *matcher = [DPLRouteMatcher matcherWithRoute:route];
+        deepLink = [matcher deepLinkWithURL:url];
+        if (deepLink) {
+            *handler = self[route];
+            break;
+        }
     }
-    else if (class_isMetaClass(object_getClass(handler)) &&
-             [handler isSubclassOfClass:[DPLRouteHandler class]]) {
-        DPLRouteHandler *routeHandler = [[handler alloc] init];
+    return deepLink;
+}
+
 
+- (BOOL)handleRoute:(id)handler withDeepLink:(DPLDeepLink *)deepLink error:(NSError *__autoreleasing *)error {
+    DPLRouteHandler *routeHandler = [self routeHandlerForHandler:handler];
+    if (routeHandler) {
         if (![routeHandler shouldHandleDeepLink:deepLink]) {
             return NO;
         }
-        
+
         UIViewController *presentingViewController = [routeHandler viewControllerForPresentingDeepLink:deepLink];
-        UIViewController <DPLTargetViewController> *targetViewController = [routeHandler targetViewController];
-        
+        UIViewController <DPLTargetViewController> *targetViewController = [self viewControllerForHandler:routeHandler withDeepLink:deepLink];
+
         if (targetViewController) {
-            [targetViewController configureWithDeepLink:deepLink];
             [routeHandler presentTargetViewController:targetViewController inViewController:presentingViewController];
         }
         else {
-            
-            NSDictionary *userInfo = @{ NSLocalizedDescriptionKey: NSLocalizedString(@"The matched route handler does not specify a target view controller.", nil)};
+
+            NSDictionary *userInfo = @{NSLocalizedDescriptionKey : NSLocalizedString(@"The matched route handler does not specify a target view controller.", nil)};
 
             if (error) {
                 *error = [NSError errorWithDomain:DPLErrorDomain code:DPLRouteHandlerTargetNotSpecifiedError userInfo:userInfo];
             }
-            
+
             return NO;
         }
     }
-    
+    else if ([handler isKindOfClass:NSClassFromString(@"NSBlock")]) {
```

Comment:
This might cause a problem in the future if the private class name changes. Can we just create a throwaway `DPLRouteHandlerBlock` and compare its `class` to `handler`'s?